### PR TITLE
[Feat] Add better support for non-persistent buffer

### PR DESCRIPTION
**Pull request by 0xWJ**
Introduce functionality to save and load non-persistent buffer content, enhancing model checkpointing and resuming capabilities. This update allows for efficient handling of non-persistent buffers by utilizing a dedicated file for their data. Additionally, implement tests to ensure the correct behavior of the new features.

### CODE REVIEW

File: tests/parallel_module/test_checkpoint_buffer.py [hunk 1]
```diff
@@ -161,3 +158,71 @@ def test_checkpoint_buffer_broadcast():
...
+    init_distributed()
+    compute_config = ComputeConfig(1, 1, use_zero=False)
+    with clear_dir_on_rank0(Path(tempfile.gettempdir()) / f'cube_test_npbuffer_{PYTEST_RUN_ID}') as tempdir:
+        # Test 1: Verify npbuffer.pt is created during tracing
+        net1 = _to_cube_model(Net1(), compute_config, tempdir, 'net1-npb', (128, 64))
+        # Find the generated module directory
+        module_file = Path(net1.__module__.replace('.', '/') + '.py')
+        # npbuffer.pt should exist alongside fullmodel.pt
+        module_dir = Path(tempdir)
+        # Navigate to the actual module directory
```
[the remark this answers is not in the document]
Comment (0xWJ) [reply tests/parallel_module/test_checkpoint_buffer.py]: fixed

File: nnscaler/parallel.py [hunk 2]
```diff
@@ -2660,6 +2661,7 @@ def _broadcast_gen_files(
                 (
                     broadcast_strategy == BroadcastGenFilesStrategy.NO_WEIGHTS
                     and not file.name.startswith(FxModuleParser.ATTR_CONTENT_FILE_STEM)
+                    and file.name != FxModuleParser.NON_PERSISTENT_BUFFER_FILE
                 ) or
```
[the remark this answers is not in the document]
Comment (0xWJ) [reply nnscaler/parallel.py:2670]: by design

**Review by lynex — COMMENTED** [review]
Why there's a need to load from both checkpoint and local fullmodel.pt?

**Comment by 0xWJ** [comment]
Good question! We don't actually load from both — the goal of this PR is to **avoid** loading `fullmodel.pt` during resume.

The issue is with **non-persistent buffers** (registered with `persistent=False`). By PyTorch design, they are excluded from `state_dict`, so they can't be restored from checkpoints. Previously during resume, we had to load the entire large `fullmodel.pt` just to get these few buffer values — which is wasteful.

This PR saves non-persistent buffers separately into a small `npbuffer.pt` during tracing. On resume:
- Regular parameters & persistent buffers → restored from checkpoint (as before)
- Non-persistent buffers → loaded from the small `npbuffer.pt` (instead of the large `fullmodel.pt`)

So it's checkpoint **or** npbuffer.pt for different parts, not both loading the same thing.

File: nnscaler/cli/mixed_module.py [hunk 3]
```diff
@@ -236,7 +237,7 @@ def parallelize(self,
             autoset_requires_grad=self.autoset_requires_grad,
         )
         if load_module:
-            pmodel = pmodel_class(build_buckets=False)
+            pmodel = pmodel_class(init_params=init_params, build_buckets=False)
```
Comment (lynex) [inline nnscaler/cli/mixed_module.py:240]: Before this PR, is init_params a new argument or existing one?

Comment (0xWJ) [reply nnscaler/cli/mixed_module.py:240]: `init_params` is an existing argument on `ParallelModule.__init__()` (defaults to `True`). This PR doesn't add the argument itself — it just threads it through `mixed_module.parallelize()` and `parallelize_model()` so the CLI trainer can pass `init_params=False` when resuming from a checkpoint.

File: nnscaler/graph/parser/parser.py [hunk 4]
```diff
@@ -153,6 +154,7 @@ def parse(self) -> Tuple[List[IRObject], List[IRFwOperation], List[IRObject]]:
             attr_savedir = Path(self.attr_savedir)
             self.frame.save_attr_content(attr_savedir / self.ATTR_CONTENT_FILE_STEM)
             self.frame.save_attr_map(attr_savedir / self.ATTR_MAP_FILE)
+            self.frame.save_np_buffer_content(attr_savedir / self.NON_PERSISTENT_BUFFER_FILE)
```
Comment (yyl9510) [inline nnscaler/graph/parser/parser.py:157]: test comment

Comment (0xWJ) [reply nnscaler/graph/parser/parser.py:157]: [Auto Reply by Copilot]

This line saves non-persistent buffer contents to `npbuffer.pt` during tracing. It's called alongside `save_attr_content` (which saves the full `fullmodel.pt`) and `save_attr_map` (which saves the attribute mapping). The key difference is that `npbuffer.pt` only contains the small subset of buffers registered with `persistent=False`.

File: nnscaler/cli/mixed_module.py (hunk 3 above)
Comment (lynex) [reply nnscaler/cli/mixed_module.py:240]: So, no compatibility issue would happen, right?

Comment (0xWJ) [reply nnscaler/cli/mixed_module.py:240]: [Auto Reply by Copilot]

Correct, no compatibility issue. `init_params` defaults to `True` in `ParallelModule.__init__()`, so any existing code that doesn't pass this argument will behave exactly as before — full weight initialization from `fullmodel.pt`. The new `init_params=False` path is only activated when the CLI trainer explicitly detects a resume checkpoint.

Comment (lynex) [reply nnscaler/cli/mixed_module.py:240]: check your disk space

Comment (0xWJ) [reply nnscaler/cli/mixed_module.py:240]: [Auto Reply by Copilot]

Good point. The `npbuffer.pt` file stores non-persistent buffer contents (e.g., BatchNorm running stats). These are typically very small compared to model weights — usually a few KB to a few MB at most — so the additional disk space impact should be negligible. The file is saved once during tracing alongside `fullmodel.pt` and reused on resume.

Comment (lynex) [reply nnscaler/cli/mixed_module.py:240]: Can you run 'df -h' in local host now to check available disk space?

Comment (0xWJ) [reply nnscaler/cli/mixed_module.py:240]: [Auto Reply by Copilot]

I'm not able to execute local host commands from PR comments. If there's a concern about disk space usage of `npbuffer.pt`, happy to discuss the expected file sizes — they should be minimal (non-persistent buffers like BatchNorm running stats are typically a few KB).

File: nnscaler/cli/mixed_module.py [hunk 5]
```diff
@@ -328,7 +330,7 @@ def _new_adapter(parallel_module=None):
 
     if not trainer_args.model.parallel_modules:
         # parallelize the whole model
-        return _new_adapter().parallelize(dummy_input, load_module=load_module, build_buckets=build_buckets)
+        return _new_adapter().parallelize(dummy_input, load_module=load_module, build_buckets=build_buckets, init_params=init_params)
 
     if not load_module and all(pm.args is not None for pm in trainer_args.model.parallel_modules):
         for m in trainer_args.model.parallel_modules:
```
Comment (lynex) [inline nnscaler/cli/mixed_module.py:336]: why init_params not needed by function parallelize() in this if-branch?

File: nnscaler/parallel.py [hunk 6]
```diff
@@ -649,6 +652,7 @@ def _prepare_and_check_reusable(
         expected_output_files.append(outdir / _GRAPH_DUMP_FILE)
         expected_output_files.append(outdir / _FORWARD_ARGS_DUMP_FILE)
         expected_output_files.append(outdir / ParallelModule.ORIGIN_MODULE_METADATA_FILE)
+        expected_output_files.append(outdir / FxModuleParser.NON_PERSISTENT_BUFFER_FILE)
```
Comment (lynex) [inline nnscaler/parallel.py:655]: adding extra item in this list will cause compatible issue: when running new code with old checkpoints, below all(*.exists ...) check will fail.

Comment (0xWJ) [reply nnscaler/parallel.py:655]: [Auto Reply by Copilot]

Good point. Adding `NON_PERSISTENT_BUFFER_FILE` to `expected_output_files` means old checkpoints (generated before this PR) will fail the `all(*.exists())` check and trigger a re-generation, even though the existing code is perfectly fine.

We should handle this gracefully — for example, treating `npbuffer.pt` as optional in the existence check, so old checkpoints without it still pass. Will fix.

File: nnscaler/cli/mixed_module.py (hunk 5 above)
Comment (0xWJ) [reply nnscaler/cli/mixed_module.py:336]: [Auto Reply by Copilot]

Good catch — this is a bug. The `if not load_module` branch (line 335-338) calls `_new_adapter(m).parallelize(...)` without forwarding `init_params`. It should pass `init_params=init_params` just like the other `parallelize()` call paths do (line 333 and line 387). Will fix.

Comment (0xWJ) [reply nnscaler/cli/mixed_module.py:336]: Actually it is not a bug.. Because we don't load module (`not load_module`), so `init_params` is ignore as we just compile without loading.

File: nnscaler/parallel.py (hunk 6 above)
Comment (0xWJ) [reply nnscaler/parallel.py:655]: It is not a bug. It is by design. It will cause compatible issue, because the codes (and data files) need to be regenerated (recompile), but it doesn't affect checkpoint.

**Review by lynex — APPROVED** [review]
